scripts/denali-dataset/match_smiles.py

```diff
@@ -0,0 +1,205 @@
+            print(ex)
+            with open("./denali_smiles_errors.dat", "a") as error_file:
+                error_file.write(f"{key}, {smiles}, {ex}\n")
+
+
+def infer_mol_from_coordinates(
+    coordinates,
+    species,
+    smiles_ref=None,
+    coordinates_unit="angstrom",
+):
+
+    # local import
+    from simtk import unit
+    from simtk.unit import Quantity
+
+    if isinstance(coordinates_unit, str):
+        coordinates_unit = getattr(unit, coordinates_unit)
+
+    # make sure we have the coordinates
+    # in the unit system
+    coordinates = Quantity(coordinates, coordinates_unit).value_in_unit(
+        unit.angstrom  # to make openeye happy
+    )
+
+    # initialize molecule
+    mol = oechem.OEGraphMol()
+
+    if all(isinstance(symbol, str) for symbol in species):
+        [
+            mol.NewAtom(getattr(oechem, "OEElemNo_" + symbol))
+            for symbol in species
+        ]
+
+    elif all(isinstance(symbol, int) for symbol in species):
+        [
+            mol.NewAtom(
+                getattr(
+                    oechem, "OEElemNo_" + oechem.OEGetAtomicSymbol(symbol)
+                )
+            )
+            for symbol in species
+        ]
+
+    else:
+        raise RuntimeError(
+            "The species can only be all strings or all integers."
+        )
+
+    ims = oechem.oemolistream()
+    ims.SetFormat(oechem.OEFormat_SMI)
+    ims.openstring(smiles_ref)
+    ref_mol = next(ims.GetOEMols())
+
+    mol.SetCoords(coordinates.reshape([-1]))
+    mol.SetDimension(3)
+    oechem.OEDetermineConnectivity(mol)
+    oechem.OEFindRingAtomsAndBonds(mol)
+    oechem.OEPerceiveBondOrders(mol)
+
+    smiles_can = oechem.OEMolToSmiles(mol)
+    smiles_ref = oechem.OEMolToSmiles(mol)
+    if smiles_ref != smiles_can:
+        print([atom.GetAtomicNum() for atom in mol.GetAtoms()])
+        print([atom.GetAtomicNum() for atom in ref_mol.GetAtoms()])
+        tmp_mol = oechem.OEGraphMol()
+        if all(isinstance(symbol, str) for symbol in species):
+            [
+                tmp_mol.NewAtom(getattr(oechem, "OEElemNo_" + symbol))
+                for symbol in species
+            ]
+
+        elif all(isinstance(symbol, int) for symbol in species):
+            [
+                tmp_mol.NewAtom(
+                    getattr(
+                        oechem, "OEElemNo_" + oechem.OEGetAtomicSymbol(symbol)
+                    )
+                )
+                for symbol in species
+            ]
+        tmp_mol.SetCoords(coordinates.reshape([-1]))
+        tmp_mol.SetDimension(3)
+        print([(atom.GetAtomicNum(), atom.GetValence()) for atom in tmp_mol.GetAtoms()])
+        oechem.OEDetermineConnectivity(tmp_mol)
+        print([(atom.GetAtomicNum(), atom.GetValence()) for atom in tmp_mol.GetAtoms()])
+        oechem.OEFindRingAtomsAndBonds(tmp_mol)
+        print([(atom.GetAtomicNum(), atom.GetValence()) for atom in tmp_mol.GetAtoms()])
+        oechem.OEPerceiveBondOrders(tmp_mol)
```

Comment:
Do you need to explicitly [perceive aromaticity](https://docs.eyesopen.com/toolkits/python/oechemtk/aromaticity.html) after this?